sync2/p2p_test.go

```diff
@@ -156,3 +157,55 @@ func TestP2P(t *testing.T) {
 			}))
 	}
 }
+
+func TestConfigValidation(t *testing.T) {
+	for _, tc := range []struct {
+		name   string
+		cfg    func(cfg *sync2.Config)
+		expErr string
+	}{
+		{
+			name: "default",
+			cfg:  func(cfg *sync2.Config) {},
+		},
+		{
+			name: "empty",
+			cfg: func(cfg *sync2.Config) {
+				*cfg = sync2.Config{}
+			},
+			expErr: "max-send-range must be positive",
+		},
+		{
+			name: "bad toplevel config",
+			cfg: func(cfg *sync2.Config) {
+				cfg.SyncInterval = 0
+			},
+			expErr: "sync-interval must be positive",
+		},
+		{
+			name: "bad range set reconciler config",
+			cfg: func(cfg *sync2.Config) {
+				cfg.RangeSetReconcilerConfig.MaxSendRange = 0
+			},
+			expErr: "max-send-range must be positive",
+		},
+		{
+			name: "bad multi peer reconciler config",
+			cfg: func(cfg *sync2.Config) {
+				cfg.MultiPeerReconcilerConfig.MaxSyncDiff = -1
+			},
+			expErr: "max-sync-diff must be non-negative",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			cfg := sync2.DefaultConfig()
+			tc.cfg(&cfg)
+			err := cfg.Validate()
+			if tc.expErr == "" {
+				require.NoError(t, err)
+			} else {
+				require.ErrorContains(t, err, tc.expErr)
```

Comment:
NIT: We should consider to return named errors from the validation, like `ErrNegativeMaxSendRange` to avoid string comparisons and use `require.ErrorIs(t, err, ErrNegativeMaxSendRange)` instead.

Reply:
IMO a bit too much hassle given the number of possible validation errors and that it will only make this test a bit prettier, not affecting anything else.
These errors are only actually used for printing, and never are checked against via `errors.Is(...)` etc except in the test. If we had some actual code that checks for particular validation errors then of course different error types would make sense.

Reply:
It's just forward thinking 😉 think about other tests we had which use this pattern of conditional asserts to save on lines of code:

Tests should be verbose and code repetition is not an anti-pattern in testing. Tests with conditionals or other branches are a PITA to debug/fix should they become an issue in the future - just look at a lot of the tests in the `hare3` and `hare4` packages. Every time I look into a failing test it takes me like 15 minutes just to figure out what the test actually does 😅

Reply:
That's not about test verbosity. The problem here is more a bit of YAGNI applied to error matching in the actual code. If there's some chance that we might have some actual code that needs to match validation errors using `errors.Is` / `errors.As`, then of course using named errors makes sense. But really it's quite unlikely we will ever do anything except failing hard and exiting upon an invalid config field value. And I think making tests "somewhat stricter" does not really justify named errors here.

Reply:
There is for instance no test case that covers that more than one configuration field is incorrect - or rather the test case that does doesn't assert this correctly, it only checks that the first field was detected correctly. So someone could change validate to only return an error instead of joining them and the test would still pass.

Here it would also help to instead assert that `Validate` logs the issues instead of merging them into a single return value. We are already using `*observer.ObservedLogs` in other tests where we assert that specific issues are logged correctly.

Reply:
Switched to logging validation errors and log observation